src/lib.rs

```diff
@@ -223,7 +212,8 @@ impl<const N: usize, A: Alphabet<N>> Debug for CustomSymbol<N, A> {
 
 impl<const N: usize, A: Alphabet<N>> Display for CustomSymbol<N, A> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        f.write_str(&self.to_string())
+        let value: String = From::<CustomSymbol<N, A>>::from(*self);
+        f.write_str(&value)
```

Comment:
I'm not sure what you meant by avoiding allocation? Do you mean in case of calling `to_string` because Formatter and buffer gets allocated?

Rustc might be able to see that `value` is just copied to result buffer.

I do get a red squigly error on current `to_string` saying to do this though.

Reply:
> Rustc might be able to see that value is just copied to result buffer.

After investigation in godbolt, even on level 3 optimization, the compiler doesn't inline formatter.